core/arch/arm/sm/suspend.S

```diff
@@ -0,0 +1,224 @@
+
+.section .text
+FUNC cpu_suspend, :
+UNWIND( .fnstart)
+	push	{r4 - r12, lr}
+UNWIND(.save	{r4 - r12, lr})
+	mov	r5, sp
+	sub	sp, sp, #SM_PM_CTX_SIZE
+	push	{r0, r1}
+
+	mov	r2, r5
+	mov	r1, #SM_PM_CTX_SIZE
+	add	r0, sp, #8
+	blx	cpu_suspend_save
+	adr	lr, aborted
+	/* Jump to arch specific suspend*/
+	pop	{r0, pc}
+aborted:
+	/* cpu not suspended */
+	add	sp, sp, #SM_PM_CTX_SIZE
+	/* Return -1 to the caller */
+	mov	r0, #(-1)
+	pop	{r4 - r12, pc}
+UNWIND( .fnend)
+END_FUNC cpu_suspend
+
+FUNC cpu_v7_suspend, :
+UNWIND(.fnstart)
+	push	{r4 - r12, lr}
+	read_fcseidr r4
+	read_tpidruro r5
+	stmia	r0!, {r4 - r5}
+	read_dacr r4
+#ifdef CFG_WITH_LPAE
+#error "Not supported"
+#else
+	read_ttbr0 r5
+	read_ttbr1 r6
+	read_ttbcr r7
+#endif
+	read_sctlr r8
+	read_actlr r9
+	read_cpacr r10
+	read_mvbar r11
+	stmia	r0!, {r4 - r11}
+	read_prrr r4
+	read_nmrr r5
+	read_vbar r6
+	read_nsacr r7
+	stmia	r0, {r4 - r7}
+	pop	{r4 - r12, pc}
+UNWIND( .fnend)
+END_FUNC cpu_v7_suspend
+
+FUNC cpu_do_suspend, :
+UNWIND(.fnstart)
+	push	{r4, r5}
+	read_midr r4
+	ubfx	r5, r4, #4, #12
+	ldr     r4, =A7_PART_NUM
+	cmp	r5, r4
+	beq	a7_suspend
+	/*
+	 * TODO: add more Cortex A[x] check?
+	 *
+	 * A9 needs PCR/DIAG
+	 */
+	read_pcr r4
+	read_diag r5
+	stmia	r0!, {r4 - r5}
+a7_suspend:
+	pop	{r4, r5}
+	b	cpu_v7_suspend
+UNWIND( .fnend)
+END_FUNC cpu_do_suspend
+
+FUNC cpu_resume, :
+UNWIND(.fnstart)
+	cpsid	aif
+
+	adr	r0, _suspend_sp
+	ldr	r2, [r0]
+	/* Get the physical address of thread_core_local */
+	add	r0, r0, r2
+
+	/* Assmue max 4 cpus and 1 cluster */
+	read_mpidr r1
+	and	r1, r1, #0x3
+	mov_imm r2, THREAD_CORE_LOCAL_SIZE
+	mla	r0, r1, r2, r0
+
+	ldr	r0, [r0, #THREAD_CORE_LOCAL_SM_PM_CTX_PHYS]
+	/* Need to use r0!, because cpu_do_resume needs it */
+	ldmia	r0!, {sp, pc}
+UNWIND( .fnend)
+END_FUNC cpu_resume
+
+FUNC cpu_resume_after_mmu, :
+UNWIND(.fnstart)
+	mov	r0, #0
+	pop	{r4 - r12, pc}
+UNWIND( .fnend)
+END_FUNC cpu_resume_after_mmu
+
+FUNC cpu_resume_mmu, :
+UNWIND(.fnstart)
+	ldr	r3, =cpu_resume_after_mmu
+	write_tlbiall
+	isb
+	write_sctlr r0
+	dsb
+	isb
+	nop
+	nop
+	nop
+	bx	r3
+UNWIND( .fnend)
+END_FUNC cpu_resume_mmu
+
+/*
+ * Restore the registers stored when cpu_do_suspend
+ */
+FUNC cpu_do_resume, :
+UNWIND(.fnstart)
+	read_midr r4
+	ubfx	r5, r4, #4, #12
+	ldr     r4, =A7_PART_NUM
+	cmp	r5, r4
+	beq	a7_resume
+
+	/*
+	 * A9 needs PCR/DIAG
+	 */
+	ldmia   r0!, {r4 - r5}
+	write_pcr r4
+	write_diag r5
+
+a7_resume:
+	/* v7 resume */
+	mov	ip, #0
+	/* Invalidate icache to PoU */
+	write_iciallu
+	/* set reserved context */
+	write_contextidr ip
+	ldmia	r0!, {r4 - r5}
+	write_fcseidr r4
+	write_tpidruro r5
+	ldmia	r0!, {r4 - r11}
+	/* Invalidate entire TLB */
+	write_tlbiall
+	write_dacr r4
+#ifdef CFG_WITH_LPAE
+#error "Not supported -"
+#else
+	write_ttbr0 r5
+	write_ttbr1 r6
+	write_ttbcr r7
+#endif
+
+	ldmia	r0, {r4 - r7}
+	write_prrr r4
+	write_nmrr r5
+	write_vbar r6
+	write_nsacr r7
+
+	read_actlr r4
+	teq	r4, r9
+	mcrne   p15, 0, r9, c1, c0, 1
```

Comment:
Since we don't understand why we should do it, let's skip it.
Doing stuff without knowing why is like cargo cult programming, https://en.wikipedia.org/wiki/Cargo_cult_programming